### PR TITLE
'Shards' now fly off sometimes when armor resists a slash

### PR DESCRIPTION
**Pull request by carlarctg**
# About the pull request

'Shards' now fly off sometimes when armor resists a slash.

What this means is that if armor absorbs (less than 34% of raw slash damage) or softens (66%), white 'sparks' will fly off with a 33 or 15 chance respectively, the amount being bigger on absorbed hits.

Added a subtype for spark effects that uses a recolored and shortened white spark effect.

https://streamable.com/tanjav

# Explain why it's good for the game

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog

:cl:
add: 'Shards' now fly off sometimes when armor resists a slash
/:cl:

### CODE REVIEW

**Review by harryob — APPROVED** [review]
code fine

File: code/game/objects/effects/effect_system/effect_system.dm [hunk 1]
```diff
@@ -102,12 +102,16 @@ steam.start() -- spawns the effect
...
+	icon_state = "armor_shards"
 	sound_to_play = null
 	amount = 4
 
 /datum/effect_system/spark_spread
 	var/total_sparks = 0 // To stop it being spammed and lagging!
+	//Amount of sleep ticks before the sparks start moving around.
+	var/spark_sleep_delay = 5
+	// How long sparks stay on the ground after creation.
+	var/spark_duration = 2 SECONDS
```
Comment (Zonespace27) [inline code/game/objects/effects/effect_system/effect_system.dm:114]: code doc

Comment (carlarctg) [reply code/game/objects/effects/effect_system/effect_system.dm:114]: HUH??????]

File: code/game/objects/effects/effect_system/effect_system.dm [hunk 2]
```diff
@@ -85,28 +85,40 @@ steam.start() -- spawns the effect
 /obj/effect/particle_effect/sparks
 	name = "sparks"
 	icon_state = "sparks"
-	var/amount = 6
 	anchored = TRUE
 	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	var/amount = 6
+	var/sound_to_play = "sparks"
```
Comment (Zonespace27) [inline code/game/objects/effects/effect_system/effect_system.dm:91]: codedoc

File: code/game/objects/effects/effect_system/effect_system.dm [hunk 3]
```diff
@@ -85,28 +85,40 @@ steam.start() -- spawns the effect
...
 
 /obj/effect/particle_effect/sparks/New()
 	..()
-	playsound(src.loc, "sparks", 25, 1)
+	if(sound_to_play)
+		playsound(src.loc, sound_to_play, 25, 1)
 // var/turf/T = src.loc
 // if (istype(T, /turf))
 // T.hotspot_expose(1000,100)
```
Comment (Zonespace27) [inline code/game/objects/effects/effect_system/effect_system.dm:99]: these may not be yours, but removing them would be nice

File: code/game/objects/effects/effect_system/effect_system.dm [hunk 4]
```diff
@@ -85,28 +85,40 @@ steam.start() -- spawns the effect
...
-/datum/effect_system/spark_spread/set_up(n = 3, c = 0, loca)
-	if(n > 10)
-		n = 10
-	number = n
-	cardinals = c
+	/// Amount of sleep ticks before the sparks start moving around.
+	var/spark_sleep_delay = 5
+	/// How long sparks stay on the ground after creation.
+	var/spark_duration = 2 SECONDS
+	var/obj/effect/particle_effect/sparks/sparks_type = /obj/effect/particle_effect/sparks
```
Comment (Zonespace27) [inline code/game/objects/effects/effect_system/effect_system.dm:115]: code doc + why is this a casted var when it's a type that never utilizes initial() anywhere

File: code/modules/mob/living/carbon/xenomorph/attack_alien.dm [hunk 5]
```diff
@@ -156,9 +156,18 @@
 			//nice messages so people know that armor works
 			if(f_damage <= 0.34*damage)
 				to_chat(src, SPAN_WARNING("Your armor absorbs the blow!"))
+				if(prob(33))
+					visible_message(SPAN_NOTICE("[M]'s [M.slash_verb] tears off pieces off [src]'s armor!"), max_distance = 4) // cool visuals indicating the armor absorbing hits.
+					var/datum/effect_system/spark_spread/armor_shards/effect = new /datum/effect_system/spark_spread/armor_shards
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/xenomorph/attack_alien.dm:161]:
```suggestion
					var/datum/effect_system/spark_spread/armor_shards/effect = new
```

File: code/modules/mob/living/carbon/xenomorph/attack_alien.dm [hunk 6]
```diff
@@ -156,9 +156,18 @@
...
+					effect.set_up(rand(3, 4), 4, src)
+					effect.start()
 			else if(f_damage <= 0.67*damage)
 				to_chat(src, SPAN_WARNING("Your armor softens the blow!"))
-
+				if(prob(15))
+					visible_message(SPAN_NOTICE("[M]'s [M.slash_verb] swipes bits off [src]'s armor!"), max_distance = 4) // cool visuals indicating the armor absorbing hits.
+					var/datum/effect_system/spark_spread/armor_shards/effect = new /datum/effect_system/spark_spread/armor_shards
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/xenomorph/attack_alien.dm:168]:
```suggestion
					var/datum/effect_system/spark_spread/armor_shards/effect = new
```

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Comment by Agameofscones** [comment]
Cool system that improves information readability,*but* goofy ass particles. Recommending they get resprited

File: code/modules/mob/living/carbon/xenomorph/attack_alien.dm [hunk 7]
```diff
@@ -156,9 +156,18 @@
...
+					visible_message(SPAN_NOTICE("[M]'s [M.slash_verb] tears pieces off [src]'s armor!"), max_distance = 4) // cool visuals indicating the armor absorbing hits.
+					var/datum/effect_system/spark_spread/armor_shards/heavy/effect = new
+					effect.set_up(rand(1, 3), 4, src)
+					effect.start()
 			else if(f_damage <= 0.67*damage)
 				to_chat(src, SPAN_WARNING("Your armor softens the blow!"))
-
+				if(prob(20))
+					visible_message(SPAN_NOTICE("[M]'s [M.slash_verb] swipes bits off [src]'s armor!"), max_distance = 4) // cool visuals indicating the armor absorbing hits.
+					var/datum/effect_system/spark_spread/armor_shards/effect = new
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/attack_alien.dm:168]:
```suggestion
					var/datum/effect_system/spark_spread/armor_shards/effect = new()
```

**Review by harryob — CHANGES_REQUESTED** [review]
ci fail

File: code/game/objects/effects/effect_system/effect_system.dm [hunk 8]
```diff
@@ -85,28 +85,47 @@ steam.start() -- spawns the effect
...
-	cardinals = c
+/datum/effect_system/spark_spread
+	/// Total sparks. Counts the total amount, does not indicate how many to create. If above 20, stops creating sparks.
+	var/total_sparks = 0
+	/// Amount of sleep ticks before the sparks start moving around.
+	var/spark_sleep_delay = 5
+	/// How long sparks stay on the ground after creation.
+	var/spark_duration = 2 SECONDS
+	/// Type of sparks to generate.
+	var/obj/effect/particle_effect/sparks/sparks_type
```
Comment (Zonespace27) [inline code/game/objects/effects/effect_system/effect_system.dm:122]:
```suggestion
	var/obj/effect/particle_effect/sparks/sparks_type = /obj/effect/particle_effect/sparks
```
Stops flaky CI fails

**Pull request closed without merging**